### PR TITLE
Prepare for FPWD

### PR DESCRIPTION
**Pull request by marcoscaceres**
To prepare for FPWD, I rewrote the spec to make it more generalized. The spec now splits API processing and JSON processing.

### CODE REVIEW

**Comment by marcoscaceres** [comment]
Please check it carefully... as it's basically a full rewrite.

File: index.html [hunk 1]
```diff
@@ -3,449 +3,298 @@
...
+  <body data-cite="ENCODING MIMESNIFF">
+    <h1 id="title">
+      Image resource
+    </h1>
     <section id="abstract">
       <p>
-        This document describes a generalized structure to be used in APIs that
-        accept image definitions.
+        This document defines the concept of an "image resource" and a
+        corresponding a WebIDL {{ImageResource}} dictionary. Web APIs can used
```
Comment (rayankans) [inline index.html]: can used -> can use

File: index.html [hunk 2]
```diff
@@ -3,449 +3,298 @@
...
         Web applications often accept developer-provided image resources to be
-        displayed. This specification defines a generalized structure where
-        developers can provide additional metadata about an image. The user
-        agent can then use this extra information to choose an an image that is
-        best suited to display on the end-user's device or most closely matches
-        the end-user's preferences.
+        displayed outside of a HTML document (e.g., in the OS, in the browser
+        UI, etc.). This specification defines a WebIDL dictionary that
+        describes an image, and from where that image resource can be fetched
+        from. The user agent can then use this extra information to choose an
```
Comment (rayankans) [inline index.html]: remove one of the 'an's

File: index.html [hunk 3]
```diff
@@ -3,449 +3,298 @@
...
       <h2>
-        <dfn>ImageResource</dfn> and its members
+        <dfn>ImageResource</dfn> dictionary
       </h2>
       <pre class="idl">
         dictionary ImageResource {
           required USVString src;
           DOMString sizes;
-          USVString type;
+          DOMString type;
```
Comment (rayankans) [inline index.html]: The manifest spec currently defines this as a USVString. Is this change intentional?

File: index.html [hunk 4]
```diff
@@ -3,449 +3,298 @@
...
+      <section>
         <h3>
           <dfn>type</dfn> member
         </h3>
         <p>
-          The <a>type</a> member of an <a>ImageResource</a> is a <a>string</a>
-          that is a <a>valid MIME type</a>. It acts as a hint to the media type
-          of the image. The purpose of this member is to allow the user agent
-          to ignore images of media types it does not support.
+          The {{ImageResource/type}} member represents a [=MIME Type=] for the
```
Comment (rayankans) [inline index.html]: MIME Type -> image MIME Type?

File: index.html [hunk 5]
```diff
@@ -3,449 +3,298 @@
...
+      </h2>
       <p>
-        The steps for <dfn>processing <a>ImageResource</a> members</dfn> are
-        given by the following algorithm. The algorithm takes an
-        Array&lt;<a>ImageResource</a>&gt; <var>entries</var> and a <a>URL</a>
-        <var>context</var>. This algorithm returns a <a>list</a> of
-        <a>processed image resources</a>.
+        The steps to <dfn data-export="">process an `ImageResource` from an
+        API</dfn> are given by the following algorithm. The algorithm takes a
+        {ImageResource} |input:ImageResource|. It returns an [=image
```
Comment (rayankans) [inline index.html]: a {ImageResource} -> an {{ImageResource}}

File: index.html [hunk 6]
```diff
@@ -3,449 +3,298 @@
...
           <ol>
-            <li>Let <var>image</var> be a new <a>processed image resource</a>.
-            </li>
-            <li>Set <var>image</var>'s <a data-link-for=
-            "processed image resource">src</a> to the result of <a>parsing</a>
-            <var>entry</var>["src"] using <var>context</var> as the <var>base
-            URL</var>. If this fails, <a>continue</a>.
+            <li>Set |image|'s [=image resource/sizes=] to the result of parsing
+            |input|.{{ImageResource/sizes}} as if it was a [^link/sizes^]
+            attribute. If parsing fails, throw a {{TypeError}}
```
Comment (rayankans) [inline index.html]: TypeError isn't a DomException name

File: index.html [hunk 7]
```diff
@@ -3,449 +3,298 @@
...
-            "processed image resource">sizes</a> to the result of running
-            <a>processing the <code>sizes</code> member of an image</a> given
-            <var>entry</var>. If this is <code>undefined</code>,
-            <a>continue</a>.
+          </ol>
+        </li>
+        <li>If |input|'s {{ImageResource/type}} member is present and not the
+        empty string:
+          <ol>
+            <li>If {{ImageResource/type}} is not a <a>valid MIME type
```
Comment (rayankans) [inline index.html]: MIME Type -> Image MIME Type?

File: index.html [hunk 8]
```diff
@@ -3,449 +3,298 @@
...
+            string</a>, throw a {{"TypeError"}} {{DOMException}}.
```
Comment (rayankans) [inline index.html]: re: TypeError not being a valid DOMException name

File: index.html [hunk 9]
```diff
@@ -3,449 +3,298 @@
...
+    </section>
+    <section>
+      <h2>
+        Fetching an image resource
+      </h2>
+      <p>
+        The steps for <dfn data-export="">fetching an image resource</dfn> are
+        given by the following algorithm. The algorithm takes a [=image
+        resource=] |image|, and an optional [=Request=] |request|. It returns a
+        <a>Response</a>.
```
Comment (rayankans) [inline index.html]: [=Response=] for consistency, here and below

File: index.html (hunk 3 above)
Comment (marcoscaceres) [reply index.html]: Yes. I can't think of a reason  to make this would be an USVString?

File: index.html (hunk 6 above)
Comment (marcoscaceres) [reply index.html]: Good catch. Will fix all.

**Comment by marcoscaceres** [comment]
After this land, we should hit up Jake first for review, then maybe Simon Peters or other folks who know the image fetching/processing mechanics and give constructive feedback.

File: index.html [hunk 10]
```diff
@@ -3,449 +3,295 @@
...
+  <body data-cite="ENCODING MIMESNIFF">
+    <h1 id="title">
+      Image resource
+    </h1>
     <section id="abstract">
       <p>
-        This document describes a generalized structure to be used in APIs that
-        accept image definitions.
+        This document defines the concept of an "image resource" and a
+        corresponding a WebIDL {{ImageResource}} dictionary. Web APIs can use
```
Comment (rayankans) [inline index.html]: corresponding a WebIDL -> corresponding WebIDL

File: index.html [hunk 11]
```diff
@@ -3,449 +3,295 @@
...
+        </dd>
+        <dt>
+          <dfn data-dfn-for="image resource">type</dfn>
+        </dt>
+        <dd>
+          An [=image MIME type=].
+        </dd>
+      </dl>
+      <p>
+        In an API, this is represented as a {{ImageResource}} dictionary.
```
Comment (rayankans) [inline index.html]: a {{ImageResource}} -> an {{ImageResource}}

File: index.html [hunk 12]
```diff
@@ -3,449 +3,295 @@
...
+      <section>
         <h3>
           <dfn>type</dfn> member
         </h3>
         <p>
-          The <a>type</a> member of an <a>ImageResource</a> is a <a>string</a>
-          that is a <a>valid MIME type</a>. It acts as a hint to the media type
-          of the image. The purpose of this member is to allow the user agent
-          to ignore images of media types it does not support.
+          The {{ImageResource/type}} member represents an [=image MIME Typ=]
```
Comment (rayankans) [inline index.html]: [=image MIME Typ=] -> [=image MIME Type=]

File: index.html [hunk 13]
```diff
@@ -3,449 +3,295 @@
...
+        Processing an `ImageResource` from an API
+      </h2>
       <p>
-        The steps for <dfn>processing <a>ImageResource</a> members</dfn> are
-        given by the following algorithm. The algorithm takes an
-        Array&lt;<a>ImageResource</a>&gt; <var>entries</var> and a <a>URL</a>
-        <var>context</var>. This algorithm returns a <a>list</a> of
-        <a>processed image resources</a>.
+        The steps to <dfn data-export="">process an `ImageResource` from an
+        API</dfn> are given by the following algorithm. The algorithm takes a
```
Comment (rayankans) [inline index.html]: a -> an

File: index.html [hunk 14]
```diff
@@ -3,449 +3,295 @@
...
-            <li>
-              <dfn data-cite="FETCH#concept-request">Request</dfn>
-            </li>
-          </ul>
-        </li>
-        <li>[[APPMANIFEST]] defines the following terms:
-          <ul>
-            <li>
-              <a data-cite="APPMANIFEST#manifest"><dfn>manifest</dfn></a>
+        <li>If is |input| has a "sizes" property and its type is a [=string=]
```
Comment (rayankans) [inline index.html]: If is |input| -> If |input|

File: index.html [hunk 15]
```diff
@@ -3,449 +3,295 @@
...
-              <dfn data-cite="INFRA#list-contain">contains</dfn>
+      </ol>
+    </section>
+    <section>
+      <h2>
+        Fetching an image resource
+      </h2>
+      <p>
+        The steps for <dfn data-export="">fetching an image resource</dfn> are
+        given by the following algorithm. The algorithm takes a [=image
```
Comment (rayankans) [inline index.html]: a -> an

File: index.html [hunk 16]
```diff
@@ -3,449 +3,295 @@
...
+        [=Response=].
+      </p>
+      <ol>
+        <li>If |request| is `undefined`:
+          <ol>
+            <li>Set |request| to a new [=Request=].
             </li>
-            <li>
-              <dfn data-cite="INFRA#list-is-empty">is empty</dfn>
+            <li>Set [=request/URL=] is set to |image|'s [=image resource/src=].
```
Comment (rayankans) [inline index.html]: drop the 'is set' here and in the lines below

**Comment by marcoscaceres** [comment]
Thanks for doing such a thorough check @rayankans! I'll try to be more careful next time.

File: index.html [hunk 17]
```diff
@@ -3,449 +3,295 @@
...
-            </ol>
-          </li>
-          <li>Return <var>set</var>.
-          </li>
-        </ol>
+        <aside class="note">
+          When multiple {{ImageResource}} dictionaries are being used by an API
+          (e.g., in a WebIDL [=sequence=]), the specifications should indicate
+          if the order in which the dictionaries are passed is significant. As
+          a general rule, order is significant and "last one wins".
```
Comment (aarongustafson) [inline index.html]: Last one wins? I thought it was usually "first match wins."

**Comment by aarongustafson** [comment]
Lots of new syntax to learn apparently.

File: index.html (hunk 17 above)
Comment (marcoscaceres) [reply index.html]: Still unsure which one is the right one... some specs do first, other do last. We might need to check a few specs.

**Comment by marcoscaceres** [comment]
@rayankans, @aarongustafson, could you please take lead on following up with some of the folks I mentioned above? ... I'd start with Jake, as he is less likely too 💩 all over it and have constructive feedback. As we are adding a fundamental primitive, we probably also want TAG review of this pretty early on too... but more importantly, we need to check if what we are proposing will be used beyond a limited set of specs we proposed here (basically, we need WHATWG folks to buy-in or this has a high risk of failure).

**Comment by marcoscaceres** [comment]
@aarongustafson, we have a nice little guide to the shorthands:

https://github.com/w3c/respec/wiki/Shorthands-Guide

I'd also suggest subscribing to the following (it's very volume... maybe 1-2 emails per month):
https://lists.w3.org/Archives/Public/spec-prod/

We make ReSpec and BikeShed announcements there about new features, syntax, etc.

File: index.html (hunk 17 above)
Comment (aarongustafson) [reply index.html]: In HTML, `audio`, `picture`, and `video` all work this way with multiple `source` elements in play (which is pretty similar).

**Comment by aarongustafson** [comment]
> @rayankans, @aarongustafson, could you please take lead on following up with some of the folks I mentioned above? ... I'd start with Jake, as he is less likely to  all over it and have constructive feedback. As we are adding a fundamental primitive, we probably also want TAG review of this pretty early on too... but more importantly, we need to check if what we are proposing will be used beyond a limited set of specs we proposed here (basically, we need WHATWG folks to buy-in or this has a high risk of failure).

Yep, I can jump on this tomorrow (on child care duties Tuesdays & Thursdays through the end of April).

**Comment by rayankans** [comment]
I reached out to @jakearchibald yesterday, reviewing the spec is on his TODO list

**Comment by aarongustafson** [comment]
Thanks @rayankans! I can submit this to the TAG as well.

**Comment by jakearchibald** [comment]
Sorry about the delay here. Crazy ol' time. Aiming to get to this next week.

File: index.html [hunk 18]
```diff
@@ -3,449 +3,295 @@
...
         Web applications often accept developer-provided image resources to be
-        displayed. This specification defines a generalized structure where
-        developers can provide additional metadata about an image. The user
-        agent can then use this extra information to choose an an image that is
-        best suited to display on the end-user's device or most closely matches
-        the end-user's preferences.
+        displayed outside of a HTML document (e.g., in the OS, in the browser
+        UI, etc.). This specification defines a WebIDL dictionary that
+        describes an image, and from where that image resource can be fetched
+        from. The user agent can then use this extra information to choose an
```
Comment (jakearchibald) [inline index.html:71]: Probably don't need the extra 'from'.

File: index.html [hunk 19]
```diff
@@ -3,449 +3,295 @@
...
+      </dl>
+      <p>
+        In an API, this is represented as an {{ImageResource}} dictionary.
+      </p>
     </section>
-    <section>
+    <section data-dfn-for="ImageResource">
       <h2>
-        <dfn>ImageResource</dfn> and its members
+        <dfn>ImageResource</dfn> dictionary
```
Comment (jakearchibald) [inline index.html:110]: Does this mean a `ImageResource` has an `ImageResource`?

File: index.html (hunk 17 above)
Comment (jakearchibald) [reply index.html:147]: With the sizes attribute, it's also first-match that wins. Are there any counter-examples, especially related to images?

File: index.html [hunk 20]
```diff
@@ -3,449 +3,295 @@
...
-        <li>[[APPMANIFEST]] defines the following terms:
-          <ul>
-            <li>
-              <a data-cite="APPMANIFEST#manifest"><dfn>manifest</dfn></a>
+        <li>If |input| has a "sizes" property and its type is a [=string=]
+        of length greater than zero:
+          <ol>
+            <li>Set |image|'s [=image resource/sizes=] to the result of parsing
+            |input|'s "sizes" property value as if it was a [^link/sizes^]
+            attribute. If parsing fails, return failure.
```
Comment (jakearchibald) [inline index.html:249]: There's a lot of repetition of the previous algorithm here. Any way to avoid it?

File: index.html [hunk 21]
```diff
@@ -3,449 +3,295 @@
...
-              <dfn data-cite="INFRA#list-append" data-lt=
-              "list-append">append</dfn>
+            <li>Set [=request/Destination=] to "image".
             </li>
-            <li>
-              <dfn data-cite="INFRA#set">set</dfn>
+            <li>Set [=request/Mode=] to "no-cors".
             </li>
-          </ul>
+          </ol>
```
Comment (jakearchibald) [inline index.html:289]: Feels like the client and credentials mode of this request should be set?

File: index.html (hunk 17 above)
Comment (aarongustafson) [reply index.html:147]: Not that I am aware of. I think we need to change it to "first match wins."

File: index.html (hunk 18 above)
Comment (aarongustafson) [reply index.html:71]: Resolved in #20

File: index.html (hunk 17 above)
Comment (aarongustafson) [reply index.html:147]: Resolved in #20

File: index.html (hunk 21 above)
Comment (jakearchibald) [reply index.html:289]: Fwiw, see the icon fetching in https://wicg.github.io/background-fetch/#background-fetch-display-algorithm

Comment (rayankans) [reply index.html:289]: Different specs tend to use different Request parameters to fetch the icon, which is why the algorithm accepts a custom request. For example there is little overlap between how BGF fetches the icon and how [WebAppManifest](https://www.w3.org/TR/appmanifest/#fetching-image-resources) handles it.

Comment (marcoscaceres) [reply index.html:289]: I wonder if we should provide some guidance there? Or set some defaults if they haven't been set?

Comment (rayankans) [reply index.html:289]: It currently defaults to the Request object's defaults plus the following:
- src -> set the the processed image resource's src
- destination -> "image"
- mode -> "no cors"

That felt like minimum amount of information needed that doesn't make assumptions on the API's behalf.

Comment (marcoscaceres) [reply index.html:289]: thanks @rayankans. I defer to @jakearchibald.
